### PR TITLE
Improve handling of feedback emitters and fix some minor bugs

### PR DESCRIPTION
**Pull request by martin-fleck-at**
#### What it does
- Provide dedicated feedback emitter object as stable emitter
- Allow cleanup functions to be calculated as a lazy function
- Log if a potentially unstable feedback emitter is used
- Split feedback in change bounds tool into dedicated files
- Properly use feedback emitter to cleanup feedback correctly

Fixes https://github.com/eclipse-glsp/glsp/issues/1323 Fixes https://github.com/eclipse-glsp/glsp/issues/1243

- Ensure correct snapping of ghost element position on grid

Fixes https://github.com/eclipse-glsp/glsp/issues/1256

#### How to test

- Check the current master and this PR to compare feedback handling.
- For particular issues, check the issue description.

#### Changelog

-   [x] This PR should be mentioned in the changelog
-   [x] This PR introduces a breaking change (if yes, provide more details below for the changelog and the migration guide)

Feedback is strongly tied to an emitter that serves as a key for the particular actions that make up the feedback. While any object can serve as and emitter, feedback de-registration and cleanup depend on it.  As such it is very important that the emitter is a stable object and not, for instance, a model element which changes with each update. While previously, we often used the tool or mouse listener themselves as emitters, this change introduces a dedicated feedback emitter object that encapsulates a set of feedback and cleanup actions and provides convenience methods to build up that feedback gradually. This is especially useful for providing hooks to add additional feedback or for tools that provide several feedbacks that are not cleaned and disposed together. Our existing tools and feedback listeners have been updated to use those emitter objects as a best practice.

### CODE REVIEW

**Comment by tortmayr** [comment]
@martin-fleck-at Could you please rebase on latest master?

**Comment by martin-fleck-at** [comment]
@tortmayr Done!

**Review by tortmayr — CHANGES_REQUESTED** [review]
Thanks @martin-fleck-at. Amazing change that really improves the Feedback API and especially working with and using emitters.

Code looks good to me (apart from a few minor comments). 
I can confirm that all mentioned issues are fixed with this change.

I also did some manual testing and tried to cover all affected tools and found a couple of issues.

- **Resize handles are no longer cleared when selecting a tool**
  On master if  I select a tool, currently active resize handles are cleared. With this change this is no longer the case. This is especially confusing when doing something like edge creation
  
- **On elemment creation there is now a visible flicker after conforming the creation with a mouse click**
  On master this is a smoother transition. I assume that we now remove the ghost feedback as part of the tool disposal. This means we have a short time period between tool disposal and server update where the "node is removed" => hence the flickering.
 - T**asknodes resizing below min width/height results in move.** 
  This issue is also present on master, so we can handle it in a follow-up/separate PR
  
  Other discovered issues, not directly related to your change:
  
  - Reconnection tool is flaky in general. Confirming a reconnect target sometimes requires multiple clicks. Sometimes the edge is not reconnected even when clicking on a valid target
  - Creating a new category node logs an error: attributes.ts:33 Error: <rect> attribute width: A negative value is not valid. ("-1")

- [a11y] Keyboard resize: no client feedback => label only moves after server update

**Comment by tortmayr** [comment]
Side note: even though this PR technically does not break anything of the core API there are a lot of breaking changes in the tools itself. We should probably mention them at least in the migration guide. Could you maybe add a short summary to the Changelog PR description? This makes it easier for us to adapt the migration guide later on.

File: packages/client/src/features/change-bounds/position-snapper.ts [hunk 1]
```diff
@@ -26,7 +26,7 @@ export class PositionSnapper {
     ) {}
 
     snapPosition(position: Point, element: GModelElement, isSnap: boolean = true): Point {
-        return isSnap && this.snapper ? this.snapper.snap(position, element) : { x: position.x, y: position.y };
+        return isSnap && this.snapper ? this.snapper.snap(position, element) : { ...position };
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply packages/client/src/features/change-bounds/position-snapper.ts:29]: To be honest, at this point I do not remember anymore. Maybe I thought that if you handed in something that was an extension of point, you'd get that back as well but it is a minor thing that is not used anywhere as far as I can tell, so I'll restore the original behavior.

File: packages/client/src/utils/gmodel-util.ts [hunk 2]
```diff
@@ -373,3 +387,21 @@ export function getDescendantIds(element?: GModelElement, skip?: (t: GModelEleme
     }
     return ids;
 }
+
+export function isNotDescendantOfAnyElement<T extends GModelElement>(elements: FluentIterable<T>): (element: T) => boolean {
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply packages/client/src/utils/gmodel-util.ts:391]: Sure, will do.

File: packages/client/src/utils/gmodel-util.ts [hunk 3]
```diff
@@ -373,3 +387,21 @@ export function getDescendantIds(element?: GModelElement, skip?: (t: GModelEleme
...
+            parent = parent.parent;
+            if (elementsSet.has(parent)) {
+                return false;
+            }
+        }
+        return true;
+    };
+}
+
+export function removeDescendants<T extends GModelElement>(elements: FluentIterable<T>): FluentIterable<T> {
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply packages/client/src/utils/gmodel-util.ts:405]: Sure, will do.

**Comment by martin-fleck-at** [comment]
Thank you very much for your feedback, Tobias! I pushed an update that should address your concerns for the most part.

> * **Resize handles are no longer cleared when selecting a tool**

This is now fixed and I also make sure we do not re-create the handles if we already have them.

> * **On element creation there is now a visible flicker after conforming the creation with a mouse click**

The reason this was a smoother transition on master is because it simply got never disposed, i.e., if you started a node creation but hit 'Escape' the ghost stayed ;-) However, I think I managed to find a fix for it, please have another look.

> * **Tasknodes resizing below min width/height results in move.**

I'm re-working the change bounds handling at the moment and I'll also have a look at that but if it is okay with you I wouldn't do it as part of this change.

**Review by tortmayr — APPROVED** [review]
Thanks for the fast update. LGTM!

**Pull request merged**